Add --config-file argument to use alternate config file at runtime

samuelgregorovic opened the pull request:
Allows user to specify alternate config file for given session via `--config-file` argument. Fixes [#1170](https://github.com/jrnl-org/jrnl/issues/1170)

samuelgregorovic committed:
added new CLI argument option --config-file

samuelgregorovic committed:
pass argument and fetch alt config file if specified

samuelgregorovic committed:
argparse argument setting update

samuelgregorovic committed:
argument alias --cf added

samuelgregorovic committed:
documentation update - usage of CLI argument

samuelgregorovic committed:
fixed name-clash + unit tests

samuelgregorovic committed:
feature test added

samuelgregorovic committed:
#1170-alternate-config-file: Auto stash before rebase of "refs/heads/#1170-alternate-config-file"

wren commented:
Thank you for submitting this! We'll take a look at it on Saturday this week.

micahellison reviewed (changes requested):
This looks great. Thanks for submitting this, and for adding in documentation along with the improvement.

Would you be willing to add a couple BDD tests to ensure support for multiple journals? I see that it's currently working, but would like to be sure that it stays working as people modify this in the future. Basically, it would confirm that jrnl can read and write to a non-default journal while using the alternative config file argument:
```
jrnl secondjournal --cf test_config.yaml new journal entry
jrnl secondjournal --cf test_config.yaml -1
```

wren commented on docs/advanced.md:
```diff
@@ -85,6 +85,25 @@ jrnl --config-override display_format fancy --config-override linewrap 20 \
 
+### Using an alternate config
+
+You can specify an alternate configuration file fot the current instance of `jrnl` using `--config-file CONFIG_FILE_PATH` where
```
```suggestion
You can specify an alternate configuration file for the current instance of `jrnl` using `--config-file CONFIG_FILE_PATH` where
```

samuelgregorovic committed:
Update docs/advanced.md

Co-authored-by: Jonathan Wren <jonathan@nowandwren.com>

samuelgregorovic commented:
> This looks great. Thanks for submitting this, and for adding in documentation along with the improvement.
> 
> Would you be willing to add a couple BDD tests to ensure support for multiple journals? I see that it's currently working, but would like to be sure that it stays working as people modify this in the future. Basically, it would confirm that jrnl can read and write to a non-default journal while using the alternative config file argument:
> ```
> jrnl secondjournal --cf test_config.yaml new journal entry
> jrnl secondjournal --cf test_config.yaml -1
> ```

i dont really have experience with BDD but i can try to do it, if you won't suggest otherwisw

micahellison commented:
> i dont really have experience with BDD but i can try to do it

Feel free to take a look at `\tests\bdd\features\multiple_journals.feature` on the develop branch. You should be able to use roughly the same syntax as the tests in there to add in this use case.

wren commented:
@samuelgregorovic If you're up for it, please give it a shot! And if it doesn't work out or you need any help, just let us know and we can jump in.

samuelgregorovic commented:
> @samuelgregorovic If you're up for it, please give it a shot! And if it doesn't work out or you need any help, just let us know and we can jump in.

deal!

samuelgregorovic commented:
> > i dont really have experience with BDD but i can try to do it
> 
> Feel free to take a look at `\tests\bdd\features\multiple_journals.feature` on the develop branch. You should be able to use roughly the same syntax as the tests in there to add in this use case.

thanks for the heads up!

samuelgregorovic committed:
BDD tests added

samuelgregorovic commented:
@wren @micahellison latest commit contains the BDD tests as you wanted. could you please look at them if they satisfy your expectations? thanks

wren commented:
@samuelgregorovic This is looking great. I think we didn't quite clarify one important point, though.

We currently have two BDD frameworks in place:
1. **Behave** (in `/features` folder)
  We're currently phasing out Behave, so we don't want to update or fix anything in the `/features` folder anymore.

2. **Pytest BDD** (in `/tests/bdd` folder)
  We recently moved over all of the tests from Behave into this in order to centralize all of our tests into Pytest (which we use for unit tests). Pytest BDD will be our BDD testing framework for the forseeable future.

Sorry for the misunderstanding! The work you've done here looks great, though. It seems like you took a current feature file, copied it over, and changed the tests to work for your tests. If you do the same thing, but start with the file mentioned above (`/tests/bdd/features/multiple_journals.feature`), then I think you'll be good to go.

The only bit of feedback I have is try to use existing config files if at all possible, rather than adding any new ones (they're kind of a pain to maintain, and we're considering getting rid of them in favor of Pytest fixtures in the future).

wren commented:
Oh, also, there seems to be a merge conflict in `tests/test_parse_args.py`. I'm not sure what the conflict is, but you can resolve it if you merge the `develop` branch into your PR.

samuelgregorovic commented:
@wren sorry for the misunderstanding. i will take a look

samuelgregorovic committed:
Merge branch 'develop' of https://github.com/jrnl-org/jrnl into jrnl-org-develop

samuelgregorovic committed:
Merge branch 'jrnl-org-develop' into #1170-alternate-config-file

wren commented:
@samuelgregorovic No worries!

Also, the develop branch had broken docs tests, which is affecting this PR. If you merge in the develop branch, it should fix that for you. Sorry!

samuelgregorovic committed:
Merge pull request #2 from samuelgregorovic/develop

fix docs conflict

micahellison commented:
Hi there @samuelgregorovic, would you like any help with the pytest-bdd tests? If you're not working on them, we can add them in.

samuelgregorovic commented:
> Hi there @samuelgregorovic, would you like any help with the pytest-bdd tests? If you're not working on them, we can add them in.

im sorry i was on a holiday and i left it unfinished. i will do it in the following days

wren commented:
> > Hi there @samuelgregorovic, would you like any help with the pytest-bdd tests? If you're not working on them, we can add them in.
> 
> im sorry i was on a holiday and i left it unfinished. i will do it in the following days

No worries! This is all volunteer work, and it's always absolutely fine too take your time. Thanks for the response, and please let us know whenever you need any help.

micahellison committed:
Begin migrating/rewording --cf tests in pytest-bdd.

Uses current directory instead of deep directory structure, but requires a given for each config file referenced

micahellison commented:
Hi there @samuelgregorovic -- I started the process of migrating the tests to pytest/bdd and pushed it to this PR, but partway through the process, I found a bug related to the upgrade for older configs. Basically, if you reference the `multiple.yaml` file in the config using `--cf`, it will upgrade it and save over your *main* config file.

If you want to see it in action, backup your jrnl config (or commit it in its own git repo) before running this from the jrnl root:
```
poetry run jrnl --cf tests/data/configs/multiple.yaml this is a test
```

micahellison committed:
Merge branch 'develop' into #1170-alternate-config-file

micahellison committed:
Fix issue where specifying a config-file that needs to be upgraded ended up upgrading the user config file instead

micahellison committed:
Uncomment and rework remaining tests for pytest-bdd instead of behave

micahellison committed:
Fix copytree for Python 3.7 (which doesn't support dirs_exist_ok)

micahellison committed:
Minor fixes to alternative config examples

micahellison commented:
I fixed that issue with overwriting the default config, then updated the rest of the tests for pytest-bdd. Had to deal with some Python 3.7 issues and a weird timeout with pa11y that somehow resolved after I cleaned up a couple small errors in the documentation. I think this is ready for review now.

micahellison committed:
Remove behave tests (behave is no longer in use)

micahellison committed:
Move config file unit test to unit test dir and use pytext path fixture instead of current directory to find test data

micahellison committed:
Use explicit "given the config exists" for copying config files instead of shoehorning in "given we use the config" twice

micahellison committed:
Change when/when to when/and

micahellison committed:
Clarify scenarios and fix indentation

micahellison committed:
Confirm primary config file isn't modified when encrypting/decrypting a journal in an alternate config file

wren commented on tests/lib/given_steps.py:
```diff
@@ -93,7 +93,10 @@ def we_use_the_config(config_file, temp_dir, working_dir):
     journal_source = os.path.join(working_dir, "data", "journals")
     journal_dest = os.path.join(temp_dir.name, "features", "journals")
-    shutil.copytree(journal_source, journal_dest)
+    try:
```
Do we still need this?

micahellison committed:
Remove try/except on copytree since I'm no longer using the same

wren reviewed (approved):
🎸
